tox.ini

```diff
@@ -1,5 +1,5 @@
 [tox]
-envlist = {py27,py34,py35,py36}-{oldest,pinned,latest,extra}-lib, {py27,py34,py35,py36}-mastereventlet, py36-branchcoverage-lib, {py27,py36}-examples, docs, {py27,py36}-static
+envlist = {py27,py34,py35,py36}-{oldest,pinned,latest,extra,gevent}-lib, {py27,py34,py35,py36}-mastereventlet, {py27,py36}-gevent-examples, py36-branchcoverage-lib, {py27,py36}-examples, docs, {py27,py36}-static
```

Comment:
First, I'd say that I should clean this up so that "gevent" isn't part of the dependency dimension but have a concurrency backend dimension (as you wrote). 

Perhaps something like:
```
envlist = {py27,py36}-static,{py27,py36}-pinned-{eventlet-gevent}-lib,{py27,py36}-{eventlet,gevent}-examples,docs,{py35,py36,py27}-{oldest,latest,extra,master}-{eventlet,gevent}-lib,py36-branchcoverage-eventlet-lib
```
?

As for the matrix itself and how it is run in tests:

If we use the current model of setting the TOXENV in travis I think we can filter and just run subsets of the dimensional matrix as we want (we already skip a subset of the total matrix).
  
```yaml
# in .travis.yml
matrix:
  include:
    - stage: quick_static2
      python: 2.7
      env: TOX_ENV=py27-static
    - stage: quick_static3
      python: 3.6
      env: TOX_ENV=py36-static
...
```

The only problem with tox defining the complete matrix anyway is if someone tries to run tox standalone (I'm not even sure if that would work). 

If we DID want to run the entire matrix and define the "subset matrix" directly in tox, we could probably simplify the travis.yml file by using: https://github.com/tox-dev/tox-travis,
but then I don't know what order they run (since we do want the order and the stages to be ordered in a certain way).

Reply:
I simplified the envlist to make the dimensions clearer:

```
{python version}-{dependency state}-{concurrency backend}-{test kind}
```
if a dimension does not exist, I leave it out.  This results in:

```
envlist =
  {py27,py36}-static,
  {py27,py35,py36}-{oldest,pinned,master}-{eventlet-gevent}-lib,
  {py35,py36}-extra-{eventlet-gevent}-lib,
  {py27,py36}-{eventlet,gevent}-examples,
  docs,
  py36-branchcoverage-eventlet-lib
```

which expands to:

```
tox -l

py27-static
py36-static
py27-oldest-eventlet-gevent-lib
py27-pinned-eventlet-gevent-lib
py27-master-eventlet-gevent-lib
py35-oldest-eventlet-gevent-lib
py35-pinned-eventlet-gevent-lib
py35-master-eventlet-gevent-lib
py36-oldest-eventlet-gevent-lib
py36-pinned-eventlet-gevent-lib
py36-master-eventlet-gevent-lib
py35-extra-eventlet-gevent-lib
py36-extra-eventlet-gevent-lib
py27-eventlet-examples
py27-gevent-examples
py36-eventlet-examples
py36-gevent-examples
docs
py36-branchcoverage-eventlet-lib
```

In travis we do not let ALL of these run (in particular I'm not testing 2.7 as much for gevent), but this can easily be changed to whatever you think is best.

Reply:
This is great. Letting tox build a full matrix and running it selectively in travis is a great solution. The matrix is also very easy to read now